cpp/src/parquet/kms_client.h

```diff
@@ -0,0 +1,81 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "parquet/exception.h"
+#include "parquet/platform.h"
+
+namespace parquet {
+namespace encryption {
+
+class PARQUET_EXPORT KeyAccessToken {
+ public:
+  KeyAccessToken() = default;
+
+  explicit KeyAccessToken(const std::string value) : value_(value) {}
+
+  void Refresh(const std::string& new_value) { value_ = new_value; }
+
+  const std::string& value() const { return value_; }
+
+  void SetDefaultIfEmpty();
+
+ private:
+  std::string value_;
+};
+
+struct PARQUET_EXPORT KmsConnectionConfig {
+  std::string kms_instance_id;
+  std::string kms_instance_url;
+  std::shared_ptr<KeyAccessToken> refreshable_key_access_token;
+  std::unordered_map<std::string, std::string> custom_kms_conf;
+
+  const std::string& key_access_token() const {
+    if (refreshable_key_access_token == NULL ||
+        refreshable_key_access_token->value().empty()) {
+      throw ParquetException("key access token is not set!");
+    }
+    return refreshable_key_access_token->value();
+  }
+
+  void SetDefaultIfEmpty();
+};
+
+class PARQUET_EXPORT KmsClient {
+ public:
+  static constexpr char KMS_INSTANCE_ID_DEFAULT[] = "DEFAULT";
+  static constexpr char KMS_INSTANCE_URL_DEFAULT[] = "DEFAULT";
+  static constexpr char KEY_ACCESS_TOKEN_DEFAULT[] = "DEFAULT";
+
+  // Wraps a key - encrypts it with the master key, encodes the result
+  // and potentially adds a KMS-specific metadata.
+  virtual std::string WrapKey(const std::string& key_bytes,
+                              const std::string& master_key_identifier) = 0;
```

Comment:
It might be useful to wrap master_key_identifier in a struct, to make it clear that is is an ID and not the actual key.

Reply:
@ggershinsky I always see `master_key_identifier` goes with a `key`, so should we pass an input param `const KeyWithMasterId& key_with_master_id` instead of 2 input string params?

Reply:
@thamht4190 this is a public interface, so better off with standard string params. KeyWithMasterId is an auxiliary class for internal usage. Also, there is a value of keeping the C++ public API identical to the Java public API, to make the interop easier in the future.

Reply:
@ggershinsky what type of interop are you thinking of?

Reply:
e.g., when data is written by parquet-cpp, and read by parquet-mr, both nodes talk to the same KMS that stores the master keys. Therefore, two versions of a custom KmsClient need to be developed, in C++ and Java. This is easier to design and implement if the KmsClient API is identical/similar in both language bindings.  So far, we've managed to keep all Parquet encryption API to be as similar as possible in Java and C++.

Reply:
I haven't reviewed the other encryption code (or the java portion of this code but I would have the same feedback on the Java API if almost all of the API was passing around strings that have different semantic meanings.  It is easy to confuse order of parameters, etc.  

I've seen not great results cargo-culting between Java and C++.  Unless this is for the express purposes of JNI don't think should necessarily be a goal.

in this case with only two parameters it might be OK.  But it seems like it is potentially easy to confuse.